### ISSUE TITLE
[Chore] Improve API endpoints in the client application

### ISSUE THREAD

**Issue by olivierobert** [pakaponk/gsr-scraping#18]
## Issue

The Next.JS-based client application interacts with the API using `fetch` and all endpoints are defined in a single file https://github.com/pakaponk/gsr-scraping/blob/main/apps/web/src/api.ts. As a result, the file is quite long (lots of scrolling involved), there is quite a bit of repetition, e.g. `headers`, request options, etc, and there is no unified/consistent error handling. 

Note that any error on the API side is not handled gracefully on the client application 😅 

![image](https://user-images.githubusercontent.com/696529/197110722-dac73c90-7987-4484-a9ad-bcdc73f3786c.png)

The following happens when logging with an invalid user account.

## Expected

What would think of refactoring the requests by:
1. Having separate files per resource (auth, user, report to mimic the API application)
2. Using Axios to make requests. Creating a utility method to send requests and leverage Axios's interceptors could remove a lot of duplicated code. Consistent error logging could be added too.

**Comment by pakaponk** [comment]
Thanks for the suggestion. 
I agreed that currently, every "fetch" function contains some common configurations and the error is not handled well enough, there is no clear difference between the `fetch` error and the error response from our API.

Regarding the first point, I do agree that separating API into modules, similar to how we separate them in Nest.js, is a better approach especially when our API continues to grow. 
For now, putting them all in `api.ts` is still tolerable for me. 
Another reason that I would like to put a hold on this is that I plan to change the API to GraphQL and leverage the GraphQL Code Generator and React Query to generate hooks for each query/mutation. This way, we can share the type between the server and the client, and we don't have to add another boilerplate code on the client side every time we add a new API.

For the second point, in my opinion, I think we can still achieve the same behavior without using Axios. To minimize the size of the bundle, I would like to avoid adding another dependency to the code if it can also be done with small native codes.
(I don't mind using Axios though if it is the standard library our team uses to handle HTTP Requests.)

What do you think about this

```js
class FetchError extends Error {}

interface ResponseErrorData {
  statusCode: number;
  message: string;
}
class ResponseError<TRawError extends ResponseErrorData> extends Error {
  status: number;
  rawError: TRawError;

  constructor(error: TRawError) {
    super(error.message);
    this.status = error.statusCode;
    this.rawError = error;
  }
}

async function extendedFetch<
  TData = any,
  TError extends ResponseErrorData = ResponseErrorData
>(url: string, options: RequestInit) {
  const { headers = {}, ...restOptions } = options;

  let res;
  try {
    res = await fetch(url, {
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      credentials: 'include',
      ...restOptions,
    });
  } catch (error) {
    if (error instanceof Error) {
      throw new FetchError(error.message);
    } else {
      throw error;
    }
  }

  if (res.ok) {
    const data = (await res.json()) as TData;
    return data as TData;
  } else {
    const error = (await res.json()) as TError;
    return new ResponseError(error);
  }
}
``` 

In summary,
1. Add common configurations, the `headers` and `credentials`, to the `extendedFetch`.
2. Throw different kinds of errors for `fetch` (`FetchError`) and Error response for our API (`ResponseError`).
3.  Add Generic to allow us to appropriately declare a type definition for the `data` and `error` from API.

**Comment by olivierobert** [comment]
The proposed solution would indeed work and be sufficient for this project. `axios` is not a requirement at all, it just avoids recreating the same functionality or missing some functionality, such as [interceptors](https://axios-http.com/docs/interceptors). Having a utility `extendedFetch` would significantly improve the current implementation (even though typing `TData` with something else that `any` would be great 😉 ).
